### DISCUSSION TITLE
Support for `Escape` key in `SelectionPrompt` and `MultiSelectionPrompt`

### DISCUSSION THREAD

**Discussion by jariq** [spectreconsole/spectre.console#700] [Ideas]
Let's talk about `SelectionPrompt` in its simplest form e.g.

```csharp
var selection = AnsiConsole.Prompt(
	new SelectionPrompt<string>()
	.AddChoices(new string[] { "a", "b", "c", "d" }));
```

User can use `Up` and `Down` arrows to change selected item and `Spacebar` or `Enter` to confirm the selection.  
However in my application I'm missing an option to cancel the selection entirely e.g by pressing `Escape` key.  
As a workaround I'm adding a special item at the beginning or at the end of the list and I handle its selection in a special way.  
However in longer lists this first requires user to get to the top or bottom of the list by pressing `Home` or `End` key.  
I believe `Escape` key would be more convenient and most importantly it would help consumer code to be shorter and easier to understand.

So I would like to add support for `Escape` key to `SelectionPrompt` and `MultiSelectionPrompt`.  
Basic idea is that prompt ends when user presses `Escape` key and it somehow indicates to the caller it was cancelled (e.g. by returning `null`, setting some new property to `true`, throwing some specific exception etc.).

Is this a feature that you would be willing to merge to Spectre.Console?

**Comment by jariq** [comment]
I just submitted one possible solution as https://github.com/spectreconsole/spectre.console/pull/711

**Reply by patriksvensson** [reply]
Sorry for not getting back sooner about this. We've all been pretty swamped, but we will try to get around to reviewing it this weekend.

**Reply by jariq** [reply]
Thanks, but there's no need to hurry. I'm now developing my app with a local copy of `spectre.console` which includes proposed changes so I'm not blocked on this anymore.

**Comment by abpiskunov** [comment]
Any progress here? We also interested in this functionality.

**Comment by jariq** [comment]
I'm closing this discussion as I'm no longer interested in working on this feature.

**Comment by ZelAnton** [comment]
```csharp
var console = new EscapeCancellableConsole(AnsiConsole.Console);

string? selected;
try
{
     selected = await console.PromptAsync(selectionPrompt);
}
catch (OperationCanceledException)
{
    selected = null;
}

// Processing selected...

exit;

sealed class EscapeCancellableConsole : IAnsiConsole
{
    readonly EscapeCancellableInput input;
    readonly IAnsiConsole console;

    public EscapeCancellableConsole(IAnsiConsole console)
    {
        this.console = console;
        input = new EscapeCancellableInput(console.Input, EscapeCancellationTokenSource);
    }

    public CancellationTokenSource EscapeCancellationTokenSource { get; } = new();

    public Profile Profile => console.Profile;
    public IAnsiConsoleCursor Cursor => console.Cursor;
    public IAnsiConsoleInput Input => input;
    public IExclusivityMode ExclusivityMode => console.ExclusivityMode;
    public RenderPipeline Pipeline => console.Pipeline;
    public void Clear(bool home) => console.Clear(home);
    public void Write(IRenderable renderable) => console.Write(renderable);

    public Task<T> PromptAsync<T>(IPrompt<T> prompt, CancellationToken cancellationToken = default)
        => AnsiConsoleExtensions.PromptAsync(this, prompt, GetMergedCancellationToken(cancellationToken));

    CancellationToken GetMergedCancellationToken(CancellationToken cancellationToken)
        => cancellationToken == CancellationToken.None || cancellationToken == EscapeCancellationTokenSource.Token
        ? EscapeCancellationTokenSource.Token
        : CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, EscapeCancellationTokenSource.Token).Token;
}

sealed class EscapeCancellableInput(IAnsiConsoleInput originalInput, CancellationTokenSource cts) : IAnsiConsoleInput
{
    public bool IsKeyAvailable() => originalInput.IsKeyAvailable();

    public ConsoleKeyInfo? ReadKey(bool intercept)
    {
        var key = originalInput.ReadKey(intercept);
        if (key?.Key == ConsoleKey.Escape)
        {
            cts.Cancel();
        }

        return key;
    }

    public async Task<ConsoleKeyInfo?> ReadKeyAsync(bool intercept, CancellationToken cancellationToken)
    {
        var key = await originalInput.ReadKeyAsync(intercept, cancellationToken);
        if (key?.Key == ConsoleKey.Escape)
        {
            await cts.CancelAsync();
        }

        return key;
    }
}
```
https://gist.github.com/ZelAnton/9b631ad409f6a2ebb6730824596a3b3d

**Reply by z00mable** [reply]
@ZelAnton Thank you. You're implementation works well. Since I use AskAsync a lot this extension is also needed in EscapeCancellableConsole class:

`public Task<T> AskAsync<T>(string prompt, CancellationToken cancellationToken = default) => AnsiConsoleExtensions.AskAsync<T>(this, prompt, GetMergedCancellationToken(cancellationToken));`
